src/core/TransformControls.tsx

```diff
@@ -112,3 +121,114 @@ export const TransformControls = React.forwardRef<TransformControlsImpl, Transfo
     ) : null
   }
 )
+
+const MultiTransformControls = React.forwardRef<TransformControlsImpl, TransformControlsProps>(
+  ({ children, domElement, onChange, onMouseDown, onMouseUp, onObjectChange, object, ...props }, ref) => {
+    const transformOnlyPropNames = [
+      'enabled',
+      'axis',
+      'mode',
+      'translationSnap',
+      'rotationSnap',
+      'scaleSnap',
+      'space',
+      'size',
+      'showX',
+      'showY',
+      'showZ',
+    ]
+
+    const { camera, ...rest } = props
+    const { enabled = true } = props
+    const transformProps = pick(rest, transformOnlyPropNames)
+    const objectProps = omit(rest, transformOnlyPropNames)
+    const gl = useThree((state) => state.gl)
+    const scene = useThree((state) => state.scene)
+    const events = useThree((state) => state.events)
+    const defaultCamera = useThree((state) => state.camera)
+    const explCamera = camera || defaultCamera
+    const explDomElement = (domElement || events.connected || gl.domElement) as HTMLElement
+    const controls = React.useMemo(
+      () =>
+        new Manipulator3D({
+          scene,
+          camera: explCamera,
+          domElement: explDomElement,
+        }),
+      [scene, explCamera, explDomElement]
+    )
+
+    const group = React.useRef<THREE.Group>()
+
+    React.useLayoutEffect(() => {
+      controls.setActive(enabled)
+    }, [controls, enabled])
+
+    React.useLayoutEffect(() => {
+      const active = controls.isActive()
+      controls.setActive(true) // Must be active to attach
+      if (object) {
+        controls.attach(object instanceof THREE.Object3D ? object : object.current)
+      } else if (group.current instanceof THREE.Object3D) {
+        controls.attach(group.current)
+      }
+      controls.setActive(active)
+      return () => void controls.detach()
+    }, [object, children, controls])
+
+    return controls ? (
+      <>
+        <primitive ref={ref} object={controls} {...transformProps} />
+        <group ref={group} {...objectProps}>
+          {children}
+        </group>
+      </>
+    ) : null
+  }
+)
+
+class Manipulator3D extends ManipulatorMesh {
+  private manipulator
+
+  constructor({ scene, camera, domElement }) {
+    const manipulator = new BaseManipulator3D(
+      scene,
+      camera,
+      {
+        domElement,
+      },
+      true
+    )
+    super(manipulator.data)
+    manipulator.mesh = this
+    this.manipulator = manipulator
+  }
+
+  addEventListener(eventName, callback) {
```

Comment:
TODO: Add support for `TransformControls` events: https://threejs.org/docs/?q=transform#examples/en/controls/TransformControls